### PR TITLE
feat: custom location provider & native user location styling

### PR DESCRIPTION
**Pull request by g4rb4g3**
## Description

Added `setCustomLocation` to MapView that allows overriding the location used for the native user location component. It takes advantage of all the functionality of mapbox-maps to animate the native user location.
Added `native user location images and scale props` that allow styling the native user location component and not only use the stock images (like the blue puck).

## Checklist

- [x] I have tested this on a device/simulator for each compatible OS
- [x] I updated the documentation with running `yarn generate` in the root folder
- [x] I added/updated a sample - if a new feature was implemented (`/example`)

## Screenshot OR Video
The red arrow with the white circle is our custom styled user location using nativeTopImage and nativeShadowImage and the animation is done with setCustomLocation on the map view.
https://github.com/rnmapbox/maps/assets/646435/650d893f-c4dd-4a05-af5f-d7af98abcbeb

### CODE REVIEW

**Comment by mfazekas** [comment]
@g4rb4g3 thanks much, looks nice, for first glance will look into it more in details

One note that instead of `topImage` I'd use `topNativeImage` to avoid confusion.  `topImage` needs to be native resource name. not a mapbox image name, which we define using the  Images component. 

We call those  nativeAssetImages there - https://github.com/rnmapbox/maps/blob/main/docs/Images.md#nativeassetimages

What do you think?

**Comment by g4rb4g3** [comment]
@mfazekas 
Glad to hear that you like it 🙂 
I have named them `nativeTopImage`, `nativeBearingImage` and `nativeShadowImage` in the `UserLocation` component.
Only in the `NativeUserLocation` I have skipped the native part since the whole component is already native and that component is not exported anyways. Is that sufficient or you still want them to be named differently?

**Comment by mfazekas** [comment]
> @mfazekas
> Glad to hear that you like it 🙂
> I have named them `nativeTopImage`, `nativeBearingImage` and `nativeShadowImage` in the `UserLocation` component.
> Only in the `NativeUserLocation` I have skipped the native part since the whole component is already native and that component is not exported anyways. Is that sufficient or you still want them to be named differently?

I'm thinking on deprecate renderMode={'native'}. And making user use NativeUserLocation instead of UserLocation if they want a native user location. Therefore `nativeTopImage` is not neccesary.

I still find 'topImage' a bit confusing as those are not mapbox image names added to style, but native asset image names.

so maybe something like:
`topImageAsset`
would clarify that. We should also update the documentation.

File: src/components/NativeUserLocation.tsx [hunk 1]
```diff
@@ -21,6 +21,26 @@ export type Props = {
    * @platform ios
    */
   iosShowsUserHeadingIndicator?: boolean;
+
+  /**
+   * The image to use as the top layer for the location indicator.
+   */
+  topImage?: string;
```
Comment (mfazekas) [inline src/components/NativeUserLocation.tsx:28]:
```suggestion
  topImageAsset?: string;
```

File: src/components/NativeUserLocation.tsx [hunk 2]
```diff
@@ -21,6 +21,26 @@ export type Props = {
    * @platform ios
    */
   iosShowsUserHeadingIndicator?: boolean;
+
+  /**
+   * The image to use as the top layer for the location indicator.
```
Comment (mfazekas) [inline src/components/NativeUserLocation.tsx:26]:
```suggestion
   * The name of native image asset to use as the top layer for the location indicator. Native asset are under Image.xcassets on iOS and the drawables directory on android
```

**Review by mfazekas — CHANGES_REQUESTED** [review]

**Comment by g4rb4g3** [comment]
@mfazekas I did a few updates to this PR and it is ready your review. 🙂

**Comment by mfazekas** [comment]
@g4rb4g3 thanks much, great work, 

I've made some changes, see #3166

Changes:
- removed custom location provider. It needs to be refactored to a component, as MapView is already a huge we don't want to add new stuff to it unless needed. Also I think we can just use interface like this: 
   ```tax
   <MapVIew>
      <CustomLocationProvider location={[lng,lat]} heading={heading} />
   </MapView>
   ```
-  topImage, bearingImage etc. is now accepts an `image` name instead of a native asset name. So it supports any image defined by the Images component.
   ```tax
   <MapView>
      <Images nativeAssetImages={['pin']}>
         <Image name={'bearing'}>
            <View style={{borderColor: 'blue', borderWidth: 2, width: 16, height: 16, borderRadius: 8, backgroundColor: 'red'}}></View>
         </Image>
      </Images>
      <NativeUserLocation topImage="pin" bearingImage="bearing" />
   </MapView>
   ```
- scale now support an expression:
    ```tsx
    <NativeUserLocation topImage="topImage" visible={true}
       scale={
            [
              "interpolate",
              ["linear"],
              ["zoom"],
              10, 1.0,
              20, 4.0
            ]
          }
          />
    ```

**Comment by g4rb4g3** [comment]
@mfazekas Great job! Thanks for putting everything together nicely. 🙂

**Pull request closed without merging**